File: leetcode_ly/java/Hash Table/ValidSudoku.java
public class Solution {
    public boolean isValidSudoku(char[][] board) {
        int m = board.length;
        int n = board[0].length;
        
        int[][] row = new int[9][9];
        int[][] col = new int[9][9];
        int[][] block = new int[9][9];
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(board[i][j] != '.'){
                    int num = board[i][j] - '1';
                    if(row[i][num] == 1 || col[num][j] == 1 || block[i / 3 * 3 + j / 3][num] == 1) return false;
                    row[i][num] = 1;
                    col[num][j] = 1;
                    block[i / 3 * 3 + j / 3][num] = 1;
                }
            }
        }
        return true;
    }
}

File: leetcode_ly/c++/HashTable/ValidSudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<int> temp(9,1);
        vector<vector<int>> row(9,temp);
        vector<vector<int>> col(9,temp);
        vector<vector<int>> block(9,temp);
        
        for(int i = 0; i < 9; i++){
            for(int j = 0; j < 9; j++){
                if(board[i][j] == '.'){
                    continue;
                }
                int k = board[i][j] - '1';
                if(row[i][k] && col[j][k] && block[i / 3 * 3 + j / 3][k]){
                    row[i][k] = col[j][k] = block[i / 3 * 3 + j / 3][k] = 0;
                }
                else{
                    return false;
                }
            }
        }
        return true;
    }
};